src/java.base/share/classes/java/lang/foreign/MemoryLayout.java

```diff
@@ -665,31 +665,34 @@ static ValueLayout valueLayout(Class<?> carrier, ByteOrder order) {
     }
 
     /**
-     * Creates a sequence layout with the given element layout and element count. If the element count has the
-     * special value {@code -1}, the element count is inferred to be the biggest possible count such that
-     * the sequence layout size does not overflow, using the following formula:
+     * Creates a sequence layout with the given element layout and element count.
      *
-     * <blockquote><pre>{@code
-     * inferredElementCount = Long.MAX_VALUE / elementLayout.bitSize();
-     * }</pre></blockquote>
-     *
-     * @param elementCount the sequence element count; if set to {@code -1}, the sequence element count is inferred.
+     * @param elementCount the sequence element count.
      * @param elementLayout the sequence element layout.
      * @return the new sequence layout with the given element layout and size.
-     * @throws IllegalArgumentException if {@code elementCount < -1}.
-     * @throws IllegalArgumentException if {@code elementCount != -1} and the computation {@code elementCount * elementLayout.bitSize()} overflows.
+     * @throws IllegalArgumentException if {@code elementCount } is negative.
      */
     static SequenceLayout sequenceLayout(long elementCount, MemoryLayout elementLayout) {
-        if (elementCount == -1) {
-            // inferred element count
-            long inferredElementCount = Long.MAX_VALUE / elementLayout.bitSize();
-            return new SequenceLayout(inferredElementCount, elementLayout);
-        } else {
-            // explicit element count
             AbstractLayout.checkSize(elementCount, true);
+            Objects.requireNonNull(elementLayout);
             return wrapOverflow(() ->
-                    new SequenceLayout(elementCount, Objects.requireNonNull(elementLayout)));
-        }
+                    new SequenceLayout(elementCount, elementLayout));
+    }
+
+    /**
+     * Creates a sequence layout with the given element layout and the maximum element
+     * count such that it does not overflow a {@code long}, using the following formula:
+     *
+     * <blockquote><pre>{@code
+     * maxElementCount = Long.MAX_VALUE / elementLayout.bitSize();
+     * }</pre></blockquote>
```

Comment:
Perhaps, now that we have another separate factory, it would be better to explain this method like we do in other cases - e.g. "this is equivalent to the following code" and the show a snippet which calls the other factory with the Long.MAX_VALUE calculation.